browser/net/brave_proxying_web_socket.cc

```diff
@@ -0,0 +1,357 @@
+  binding_as_websocket_.Bind(std::move(proxied_request));
+
+  binding_as_websocket_.set_connection_error_handler(
+      base::BindRepeating(&BraveProxyingWebSocket::OnError,
+                          base::Unretained(this), net::ERR_FAILED));
+}
+
+BraveProxyingWebSocket::~BraveProxyingWebSocket() {
+  if (ctx_) {
+    request_handler_->OnURLRequestDestroyed(ctx_);
+  }
+}
+
+// static
+bool BraveProxyingWebSocket::ProxyWebSocket(
+    content::RenderFrameHost* frame,
+    network::mojom::WebSocketRequest* request,
+    network::mojom::AuthenticationHandlerPtr* auth_handler) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  network::mojom::WebSocketPtrInfo proxied_socket_ptr_info;
+  auto proxied_request = std::move(*request);
+  *request = mojo::MakeRequest(&proxied_socket_ptr_info);
+
+  base::PostTaskWithTraits(
+      FROM_HERE, {content::BrowserThread::IO},
+      base::BindOnce(&ResourceContextData::StartProxyingWebSocket,
+                     frame->GetProcess()->GetBrowserContext()
+                         ->GetResourceContext(),
+                     frame->GetProcess()->GetID(),
+                     frame->GetRoutingID(),
+                     frame->GetFrameTreeNodeId(),
+                     frame->GetLastCommittedOrigin(),
+                     std::move(proxied_socket_ptr_info),
+                     std::move(proxied_request)));
+  return true;
+}
+
+void BraveProxyingWebSocket::AddChannelRequest(
+    const GURL& url,
+    const std::vector<std::string>& requested_protocols,
+    const GURL& site_for_cookies,
+    std::vector<network::mojom::HttpHeaderPtr> additional_headers,
+    network::mojom::WebSocketClientPtr client) {
+  if (binding_as_client_.is_bound() || !client || forwarding_client_) {
+    // Illegal request.
+    proxied_socket_ = nullptr;
+    return;
+  }
+
+  request_.url = url;
+  // TODO(iefremov): site_for_cookies is not enough, we should find a way
+  // to initialize NetworkIsolationKey.
+  request_.site_for_cookies = site_for_cookies;
+  request_.request_initiator = origin_;
+  request_.render_frame_id = frame_id_;
+  websocket_protocols_ = requested_protocols;
+  request_id_ = request_id_generator_->Generate();
+
+  forwarding_client_ = std::move(client);
+  additional_headers_ = std::move(additional_headers);
+
+  // If the header client will be used, we start the request immediately, and
+  // OnBeforeSendHeaders and OnSendHeaders will be handled there. Otherwise,
+  // send these events before the request starts.
+  base::RepeatingCallback<void(int)> continuation;
+  continuation = base::BindRepeating(
+      &BraveProxyingWebSocket::OnBeforeRequestComplete,
+      weak_factory_.GetWeakPtr());
+
+  // TODO(yhirano): Consider having throttling here (probably with aligned with
+  // WebRequestProxyingURLLoaderFactory).
+  bool should_collapse_initiator = false;
+  ctx_ = std::make_shared<brave::BraveRequestInfo>();
+  brave::BraveRequestInfo::FillCTX(request_, process_id_,
```

Comment:
We use the handcrafted `request_` here, because it is not provided by WebSocket machinery. Referer is never used with WebSockets, and other non-initialized fields are basically not needed to create a context (except the request type - I've added a TODO to `url_context.h`)